pkg/handler/enqueue.go

```diff
@@ -47,8 +48,11 @@ func (e *EnqueueRequestForObject) Create(evt event.CreateEvent, q workqueue.Rate
 
 // Update implements EventHandler
 func (e *EnqueueRequestForObject) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	// requestSet will help to deduplicate requests in a single event
+	requestSet := reconcile.NewRequestSet()
```

Comment:
```suggestion
	var namespace, name string
```

Lets simplify this by just storing these two here and adding a check inside the `if evt.ObjectNew != nil` to avoid enqueueing when namespace/name are the same (which I believe will always be the case, I don't think its possible to rename objects).

There is no need for a new type, threadsafety or extending our public api with aspects of such a tiny implementation detail here.

Reply:
@alvaroaleman  Thanks for your comment, but your simple way seems only work for `EnqueueRequestForObject` handler, in some other handlers, like `EnqueueRequestsForMapFunc` [https://github.com/kubernetes-sigs/controller-runtime/blob/e1a725df2743147795e5dfc8275365f7ada24805/pkg/handler/enqueue_mapped.go#L60-L63](https://github.com/kubernetes-sigs/controller-runtime/blob/e1a725df2743147795e5dfc8275365f7ada24805/pkg/handler/enqueue_mapped.go#L60-L63), it's hard to replace `RequeueSet` simply by two variables namespace & name because there probably will be several requests.
I provide a more general way to deduplicate requests in a single event, though it looks a little bit complicated.
But I have no objections on your suggestion, just we need to find a better way to fix this issue for **all** handlers.

Reply:
You can simply split `mapAndEnqueue` into one func that creates a slice of `types.NamespacedName` and one func that iterates and enqueues. Then just use the former in `Update` and deduplicate before enqueuing. Also again, this is an internal detail that does not need extending of our public api.

Reply:
Just a little afraid of doing some similar inline codes... :)

Reply:
Should we implement something for https://github.com/kubernetes-sigs/controller-runtime/issues/857 that provides a useful feature for folks and we can put behind a flag for the time being, instead of changing critical code paths?

Reply:
#857 is orthogonal I think, this here is about deduplicating events from a single event, #857 is about ratelimiting reconciliations, if I understand it correctly

Reply:
If I understood 857 correctly though, it's about ratelimiting events for the same resource in a set time window. It seems here we're kind of doing the same, where the time window is small?

Reply:
The difference is that 857 is across multiple events (which is something we can absolutely not do by default) and this is only deduplicating requests from the same event, which is something we can do by default.